File: database/database-annotation-processor/src/test/java/ru/tinkoff/kora/database/common/annotation/processor/r2dbc/MockR2dbcExecutor.java
package ru.tinkoff.kora.database.common.annotation.processor.r2dbc;

import io.r2dbc.spi.*;
import org.mockito.Mockito;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import ru.tinkoff.kora.database.common.telemetry.DataBaseTelemetry;
import ru.tinkoff.kora.database.r2dbc.R2dbcConnectionFactory;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@SuppressWarnings("unchecked")
public class MockR2dbcExecutor implements R2dbcConnectionFactory {
    public final Connection con = Mockito.mock(Connection.class);
    public final Statement statement = Mockito.mock(Statement.class);
    public final DataBaseTelemetry telemetry = Mockito.mock(DataBaseTelemetry.class);
    public final DataBaseTelemetry.DataBaseTelemetryContext telemetryContext = Mockito.mock(DataBaseTelemetry.DataBaseTelemetryContext.class);
    public List<List<MockColumn>> rows = new ArrayList<>();

    public MockR2dbcExecutor() {
        reset();
    }

    public void setRow(MockColumn mockColumn) {
        this.setRow(List.of(mockColumn));
    }

    public void setRow(List<MockColumn> mockColumns) {
        this.setRows(List.of(mockColumns));
    }

    public void setRows(List<List<MockColumn>> mockRows) {
        this.rows.clear();
        this.rows.addAll(mockRows);
    }

    public void reset() {
        Mockito.reset(con, statement, telemetry, telemetryContext);
        rows = new ArrayList<>();
        when(con.createStatement(any())).thenReturn(statement);
        when(statement.execute()).thenReturn((Publisher) Flux.defer(() -> Flux.just(new MockResult(this.rows, null))));
        when(telemetry.createContext(any(), any())).thenReturn(telemetryContext);
    }

    public void setUpdateCountResult(long updateCount) {
        when(statement.execute()).thenReturn((Publisher) Flux.defer(() -> Flux.just(new MockResult(null, updateCount))));
    }

    public record MockColumn(String label, Object value) {}

    @Override
    public Mono<Connection> currentConnection() {
        return null;
    }

    @Override
    public Mono<Connection> newConnection() {
        return null;
    }

    @Override
    public DataBaseTelemetry telemetry() {
        return this.telemetry;
    }

    @Override
    public <T> Mono<T> inTx(Function<Connection, Mono<T>> callback) {
        return null;
    }

    @Override
    public <T> Mono<T> withConnection(Function<Connection, Mono<T>> callback) {
        return callback.apply(con);
    }


    public <T> Mono<T> inTx(Connection connection, Function<Connection, Mono<T>> callback) {
        return callback.apply(connection);
    }

    public <T> Flux<T> withConnectionFlux(Function<Connection, Flux<T>> callback) {
        return callback.apply(con);
    }

    public <T> Flux<T> inTxFlux(Connection connection, Function<Connection, Flux<T>> callback) {
        return callback.apply(connection);
    }

    private record MockResult(@Nullable List<List<MockColumn>> rows, @Nullable Long updateCount) implements Result {

        @Override
        public Publisher<Long> getRowsUpdated() {
            return Mono.justOrEmpty(updateCount);
        }

        @Override
        public <T> Publisher<T> map(BiFunction<Row, RowMetadata, ? extends T> mappingFunction) {
            var mock = Mockito.mock(RowMetadata.class);
            when(mock.getColumnMetadata(Mockito.anyInt())).thenAnswer(invocation -> {
                var i = (Integer) invocation.getArguments()[0];
                var row = rows.get(0);
                var label = row.get(i).label;
                var meta = Mockito.mock(ColumnMetadata.class);
                when(meta.getName()).thenReturn(label);
                return meta;
            });
            return Flux.fromIterable(this.rows).map(MockRow::new).map(row -> mappingFunction.apply(row, mock));
        }

        @Override
        public Result filter(Predicate<Segment> filter) {
            return this;
        }

        @Override
        public <T> Publisher<T> flatMap(Function<Segment, ? extends Publisher<? extends T>> mappingFunction) {
            throw new IllegalStateException();
        }
    }

    private record MockRow(List<MockColumn> columns) implements Row {
        @Override
        public <T> T get(int index, Class<T> type) {
            return type.cast(columns.get(index).value);
        }

        @Override
        public <T> T get(String name, Class<T> type) {
            for (var column : columns) {
                if (column.label.equals(name)) {
                    return type.cast(column.value);
                }
            }
            throw new IllegalStateException();
        }

        @Override
        public RowMetadata getMetadata() {
            throw new IllegalStateException();
        }
    }
}

File: database/database-symbol-processor/src/test/kotlin/ru/tinkoff/kora/database/symbol/processor/r2dbc/MockR2dbcExecutor.kt
package ru.tinkoff.kora.database.symbol.processor.r2dbc

import io.r2dbc.spi.*
import org.mockito.Mockito
import org.mockito.invocation.InvocationOnMock
import org.mockito.kotlin.any
import org.reactivestreams.Publisher
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.tinkoff.kora.database.common.telemetry.DataBaseTelemetry
import ru.tinkoff.kora.database.r2dbc.R2dbcConnectionFactory
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Predicate

class MockR2dbcExecutor() : R2dbcConnectionFactory {
    val con: Connection = Mockito.mock(Connection::class.java)
    val statement: Statement = Mockito.mock(Statement::class.java)
    val telemetry = Mockito.mock(DataBaseTelemetry::class.java)!!
    val telemetryCtx = Mockito.mock(DataBaseTelemetry.DataBaseTelemetryContext::class.java)!!

    var rows = ArrayList<List<MockColumn>>()

    init {
        reset()
    }

    fun reset() {
        rows.clear()
        Mockito.reset(con, statement, telemetry, telemetryCtx)
        Mockito.`when`(telemetry.createContext(any(), any())).thenReturn(telemetryCtx)
        Mockito.`when`(con.createStatement(any())).thenReturn(statement)
        Mockito.`when`(statement.execute()).thenReturn(Flux.just(MockResult(rows)))
    }


    fun setRow(mockColumn: MockColumn) {
        this.setRow(listOf(mockColumn))
    }

    fun setRow(mockColumns: List<MockColumn>) {
        setRows(listOf(mockColumns))
    }

    fun setRows(mockRows: List<List<MockColumn>>) {
        rows.clear()
        rows = ArrayList(mockRows)
    }

    data class MockColumn(val label: String, val value: Any?)

    override fun currentConnection(): Mono<Connection> {
        return Mono.fromRunnable { this.con }
    }

    override fun newConnection(): Mono<Connection> {
        return Mono.fromRunnable { this.con }
    }

    override fun telemetry() = telemetry

    override fun <T : Any?> inTx(callback: Function<Connection, Mono<T>>?): Mono<T> {
        TODO("Not yet implemented")
    }

    override fun <T> withConnection(callback: Function<Connection, Mono<T>>): Mono<T> {
        return callback.apply(con)
    }

    override fun <T : Any?> withConnectionFlux(callback: Function<Connection, Flux<T>>?): Flux<T> {
        TODO("Not yet implemented")
    }


    private data class MockResult(val rows: List<List<MockColumn>>) : Result {
        override fun getRowsUpdated(): Publisher<Long> {
            return Mono.just(0)
        }

        override fun <T> map(mappingFunction: BiFunction<Row, RowMetadata, out T>): Publisher<T> {
            val mock = Mockito.mock(RowMetadata::class.java)
            Mockito.`when`(mock.getColumnMetadata(Mockito.anyInt())).thenAnswer { invocation: InvocationOnMock ->
                val i = invocation.arguments[0] as Int
                val row = rows[0]
                val label: String = row[i].label
                val meta = Mockito.mock(ColumnMetadata::class.java)
                Mockito.`when`(meta.name).thenReturn(label)
                meta
            }
            return Flux.fromIterable(
                rows
            ).map { columns -> MockRow(columns) }
                .map { row: MockRow -> mappingFunction.apply(row, mock) }
        }

        override fun filter(filter: Predicate<Result.Segment>): Result {
            TODO("Not yet implemented")
        }

        override fun <T : Any?> flatMap(mappingFunction: Function<Result.Segment, out Publisher<out T>>): Publisher<T> {
            TODO("Not yet implemented")
        }
    }

    private data class MockRow(val columns: List<MockColumn>) : Row {
        override fun <T> get(index: Int, type: Class<T>): T {
            if (columns[index].value is Integer) {
                return columns[index].value as T
            }
            return type.cast(columns[index].value)
        }

        override fun <T> get(name: String, type: Class<T>): T {
            for (column in columns) {
                if (column.label == name) {
                    if (column.value is Integer) {
                        return column.value as T
                    }
                    return type.cast(column.value)
                }
            }
            throw IllegalStateException()
        }

        override fun getMetadata(): RowMetadata {
            TODO("Not yet implemented")
        }
    }
}
